modules/rostests/kmtests/hal/HalPortIo.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * PROJECT:     ReactOS Kernel-Mode Tests
+ * LICENSE:     LGPL-2.1-or-later (https://spdx.org/licenses/LGPL-2.1-or-later)
+ * PURPOSE:     Tests for string I/O intrinsic functions
+ * COPYRIGHT:   Copyright 2025 Dmitry Borisov <di.sean@protonmail.com>
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <kmt_test.h>
+
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+/* Trigger a bug in the CORE-20078 */
+#ifdef _MSC_VER
+#pragma optimize("gst", on)
+#pragma auto_inline(on)
+#else
+#pragma GCC optimize("O3")
+#endif
+
+typedef struct _TEST_CONTEXT
+{
+    PVOID OldBuffer;
+    UCHAR Pad[78];
+    ULONG Port;
+    ULONG Size;
+    PVOID Buffer;
+} TEST_CONTEXT, *PTEST_CONTEXT;
+
+/* FUNCTIONS ******************************************************************/
+
+static
+DECLSPEC_NOINLINE
+VOID
+TestReadStringUchar(
+    _In_ PTEST_CONTEXT Context)
+{
+    __inbytestring(Context->Port, Context->Buffer, Context->Size);
+
+    ok_eq_pointer(Context->Buffer, Context->OldBuffer);
+}
+
+static
+DECLSPEC_NOINLINE
+VOID
+TestReadStringUshort(
+    _In_ PTEST_CONTEXT Context)
+{
+    __inwordstring(Context->Port, Context->Buffer, Context->Size / sizeof(USHORT));
+
+    ok_eq_pointer(Context->Buffer, Context->OldBuffer);
+}
+
+static
+DECLSPEC_NOINLINE
+VOID
+TestReadStringUlong(
+    _In_ PTEST_CONTEXT Context)
+{
+    __outdwordstring(Context->Port, Context->Buffer, Context->Size / sizeof(ULONG));
+
+    ok_eq_pointer(Context->Buffer, Context->OldBuffer);
+}
+
+static
+DECLSPEC_NOINLINE
+VOID
+TestWriteStringUchar(
+    _In_ PTEST_CONTEXT Context)
+{
+    __inbytestring(Context->Port, Context->Buffer, Context->Size);
+
+    ok_eq_pointer(Context->Buffer, Context->OldBuffer);
+}
+
+static
+DECLSPEC_NOINLINE
+VOID
+TestWriteStringUshort(
+    _In_ PTEST_CONTEXT Context)
+{
+    __outwordstring(Context->Port, Context->Buffer, Context->Size / sizeof(USHORT));
```

Comment:
Everywhere:
```suggestion
    Context.Buffer = Context->OldBuffer;

    __outwordstring(Context->Port, Context->Buffer, Context->Size / sizeof(USHORT));
```

Reply:
This would affect the test causing unintended outcomes.

Reply:
Sorry, could you explain with some details?
I would think setting the value before is wanted in order to check it after.

Reply:
I put a DECLSPEC_NOINLINE on the isolated functions to prevent the optimizer from changing the `ok_eq_pointer` check. There shouldn't be any initialization code around.

Reply:
Ok, then:
* Do the initialization before each function call.
* Add your explanation as a code comment?

Reply:
Good point, done.
> Do the initialization before each function call.

Well, the value doesn't change between two tests.